<a href="https://colab.research.google.com/github/j-claudinei-f/j-claudinei-f/blob/main/Recorr%C3%AAncias_ordem1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Matemática Discreta**

José Claudinei Ferreira

Universidade Federal de Alfenas

#**Relações de recorrência lineares de ordem 1**

[Link](https://github.com/j-claudinei-f/j-claudinei-f/blob/main/SomasPolinomiais.ipynb) para somas polinomiais e exponenciais, para resolução de relações de reconrência da forma
$$a_{n+1}=a_n+f(n+1),$$ com $f(n)$ e $a_1$ conhecidos.

#**Relações de recorrência lineares de ordem 1**

Agora estamos interessados nas resolução de relações de reconrência da forma
$$a_{n+1}=g(n)a_n+h(n),$$ com $g(n)$, $h(n)$ e $a_1$ conhecidos.

O método de resolução é o seguinte (Teorema 4.1 do livro texto):

1. Considere uma sequência $b_n$, solução não nula da recorrência <font color=blue>$$b_{n+1}=g(n)b_n.$$

2. Defina $y_n$ como uma sequência, tal que <font color=green> $$b_ny_n=a_n.$$

3. Segue que $$\begin{cases}\color{green}{b_{n+1}y_{n+1}}&=&\color{green}{a_{n+1}}\\&=&g(n)a_n+h(n)\\&=&g(n)\color{green}{b_n y_{n}}+h(n)\\&=&\color{blue}{g(n)b_n }y_{n}+h(n)\\&=&\color{blue}{b_{n+1}}y_n+h(n)\end{cases}$$

Considerando o início e o final da última exeperssão, temos que $$y_{n+1}=y_n+\frac{h(n)}{g(n)b_n},\qquad y_1b_1=a_1,$$  que é uma expressão mais simples de manipular, usando somatório.

#**Exemplo**

 Considere a relação de recorrência $$a_{n+1}=3 a_n+2,$$ com $a_1=2$, que vem de uma forma de calcular o número mínimos de movimentos na torre de Hanói (com a regra de sempre passar pela haste do meio). 

Calcule $a_{10}$ e determine uma expessão para $a_n$ sem recorrência.

**Uma parte da solução:**

Para calcular $a_{10}$ basta calcularmos $a_2,\,a_3,\,\ldots,\,a_9$ e $a_{10}$, usando a recorrência.

In [ ]:
a=[2]

for i in range(1,10):
  a=a+[3*a[i-1]+2]

a

[2, 8, 26, 80, 242, 728, 2186, 6560, 19682, 59048]

Isso nos dá $a_{10}=59048$.

**Uma solução geral**

Uma forma de resolver esse problema é resolver antes a recorrência
$$b_{n+1}=3b_n,$$ que tem $b_n=3^{n}$ como solução, por exemplo. 

Agora obtemos uma relação de recorrência $$y_{n+1}=y_n+\frac{2}{3^{n}\times 3},\qquad 3^{1}y_1=a_1=2.$$

Devemos ter $a_n=3^ny_n$.

Mas, $$y_2=\frac{2}{3}+\frac{2}{3}\frac{1}{3^1}=\frac{2}{3}\left(1+\frac{1}{3^1}\right)$$
$$y_3=\frac{2}{3}\left(1+\frac{1}{3^1}+\frac{1}{3^2}\right)$$
e podemos ver que 
$$y_n=\frac{2}{3}\left(1+\frac{1}{3^1}+\frac{1}{3^2}+\cdots+\frac{1}{3^{n-1}}\right)=\frac{2}{3}\left(\frac{\frac{1}{3^{n}}-1}{\frac{1}{3}-1}\right).$$

Por fim, $$a_n=3^n\frac{2}{3}\left(\frac{\frac{1}{3^{n}}-1}{\frac{1}{3}-1}\right)=3^n-1.$$

In [ ]:
3**10-1   # 3^10-1=a(10)

59048

#**Uma forma para somar produto de expressões**

Quando vimos formas de considerar as relações de recorrência
$$a_{n+1}=a_n+f(n+1),$$
começamos com uma escolha de $a_1$ e obtemos: $$a_2=a_1+f(2)=(a_1-f(1))+f(1)+f(2),$$ e continuamos até obtermos $$a_n=(a_1-f(1))+f(1)+f(2)+\cdots f(n).$$

Já vimos como calcular a soma, de forma reduzida, para os casos em que $$f(i)=\sum_{i=1}^pc_ii^m$$ é um polinômio ($m\in\mathbb{N}$) e no caso em que $$f(i)=\sum_{i=1}^pc_ir_i^i$$ é uma soma de exponenciais.

Segue que sabemos manipular o somatório $$\sum_{i=1}^n f(i),$$ no caso em que $f(i)$ é uma soma de polinômios com exeponenciais (da forma mencionada).


Entretanto, há casos em que $$f(n)=g(n)h(n)$$ é um produto e podemos usar a seguinte ideia para tentar manipular o somatório 
$$\sum_{i=1}^nf(i)=\sum_{i=1}^ng(i)h(i).$$

**O operador diferenças**

Em momento anterior da disciplina foi definido o operador diferença 
$$\Delta a_n=a_{n+1}-a_n.$$

Segue dessa definição que vale a igualdade na soma $$\sum_{i=k}^n\Delta a_i=a_{n+1}-a_k.$$

**O teorema das somas por partes:**

Considere uma sequência $a_n=b_nc_n$, em que $b_n$ e $c_n$ são sequências conhecidas. 

Então, $$\Delta a_n=a_{n+1}-a_n=b_{n+1}c_{n+1}-b_nc_n.$$

Você pode fazer um ajuste nessa expressão e escrevê-la como 
$$\Delta b_nc_n=b_{n+1}c_{n+1}-b_nc_n=(b_{n+1}c_{n+1}-b_{n+1}c_n)+(b_{n+1}c_n-b_nc_n)=b_{n+1}\Delta c_n+c_{n}\Delta b_n.\tag{$\star$}$$

#**Exemplo**

Resolver a relação de recorrência $$a_{n+1}=3a_n+n,\qquad a_1=1.$$

**Uma forma de resolução:**

Já vimos que uma solução da relação $$x_{n+1}=3x_n$$ é $x_n=3^n$. Isso nos permite escrever $a_n=3^ny_n$ e encontrar $y_n$, por meio da expressão $$y_{n+1}=y_{n}+\frac{n}{3^n\times 3},\qquad 3^1y_1=a_1.$$

Então, temos $$y_2=y_1+\frac{1}{3^2}=\frac{1}{3}+\frac{1}{3^2}$$ e  $$y_3=y_2+\frac{2}{3^3}=\frac{1}{3}+\frac{1}{3^2}+\frac{2}{3^3}$$, ....

Logo, $$y_n=\frac{1}{3}+\sum_{i=1}^{n-1}\frac{i}{3^{i+1}},\quad n>1.$$

Do que já falamos no curso, o que tem de novo é o somatório envolvendo o produto $g(n)h(n)$, em que $g(n)=n=g_n$ é um polinômio e  $h(n)=\frac{1}{3^{n+1}}=h_n$ é exponêncial.

Obseve ainda que
$$\Delta g_n=g(n+1)-g(n)=1$$
e que  $$\Delta h_n=h_{n+1}-h_n=\frac{1}{3^{n+2}}-\frac{1}{3^{n+1}}=\frac{1}{3^{n+1}}\left(\frac{1}{3}-1\right)=-\frac{1}{3^{n+1}}\frac{2}{3}=-2\times\frac{1}{3^{n+2}},$$ ou que $$-\frac{1}{2}\Delta h_n=\frac{1}{3^{n+2}},\qquad n\in \mathbb{Z}.$$

Podemos usar isso para escrever a igualdade
$$\sum_{i=1}^{n-1}\frac{i}{3^{i+1}}=\sum_{i=1}^{n-1}g_ih_i=-\sum_{i=1}^{n-1}g_i\left(\frac{1}{2}\Delta h_{i-1}\right)=-\frac{1}{2}\sum_{i=0}^{n-2}g_{i+1}\Delta h_{i}.$$

Usando agora a expressõa $(\star)$, podemos reescrever essa soma como 
$$\sum_{i=1}^{n-1}\frac{i}{3^{i+1}}=-\frac{1}{2}\sum_{i=0}^{n-2}\left(\Delta g_{i}h_i-h_i\Delta g_{i}\right)=-\frac{1}{2}\left(\sum_{i=0}^{n-2}\Delta g_{i}h_i-\sum_{i=0}^{n-2}\frac{1}{3^{i+1}}\right).$$

Por fim,
$$\sum_{i=1}^{n-1}\frac{i}{3^{i+1}}=-\frac{1}{2}\left(g_{n-1}h_{n-1}-g_0h_0-\frac{1}{3}\left(\frac{\left(\frac{1}{3})\right)^{n-1}-1}{\frac{1}{3}-1}\right)\right)=-\frac{1}{2}\left((n-1)\frac{1}{3^n}+\frac{1}{2}\left(\frac{1}{3^{n-1}}-1\right)\right).$$ E segue que 
$$a_n=3^ny_n=3^n\times \frac{1}{3}-3^n\times \frac{1}{2}\left((n-1)\frac{1}{3^n}+\frac{1}{2}\left(\frac{1}{3^{n-1}}-1\right)\right)=3^{n-1}-\frac{n-1}{2}-\frac{3}{4}+\frac{3^n}{4}.$$

Ou seja, $$a_n=\frac{7}{12}\times 3^{n}-\frac{2n+1}{4}.$$

**Testendo o resultado obtido:** 

Vamos testar o que obtemos:

Primeiro olhamos para $$a_{n+1}=3a_n+n,\qquad a_1=1,$$ e calculamos alguns $a_i$.

In [6]:
a=[1]                     # Lista com 1 termo.

for i in range(1,10):
  a=a+[3*a[i-1] + i-1]      # Aumento a lista com um termo.
a

[1, 3, 10, 32, 99, 301, 908, 2730, 8197, 24599]

Agora olhamos para 
$$a_n=\frac{7}{12}\times 3^{n}-\frac{2n+1}{4},$$ e comparamos com os valores anteriores.

In [5]:
a=[]                  # Espaço vazio para guardar números depois.
for i in range(0,10):
  a=a+[15*(3**i)/12-(2*i+1)/4]
a

[1.0, 3.0, 10.0, 32.0, 99.0, 301.0, 908.0, 2730.0, 8197.0, 24599.0]

Isso sugere que não erramos cálculos.